php/Blockonomics.php

```diff
@@ -238,4 +294,334 @@ public function testSetup()
+    }
+
+    public function create_new_order($order_id, $crypto){
+        $responseObj = $this->new_address(get_option("blockonomics_callback_secret"), $crypto);
+        if($responseObj->response_code != 200) {
+            exit();
+        }
+        $address = $responseObj->address;
+
+        $order = array(
+                'order_id'           => $order_id,
+                'status'             => -1,
+                'crypto'             => $crypto,
+                'address'            => $address
+        );
+        $order = $this->update_waiting_order($order);
+
+        $this->record_address($order_id, $crypto, $address);
+        return $order;
+    }
+
+    // Load the the checkout template in the page
+    public function load_checkout_template($order_id, $crypto){
+        // Check to send the user to nojs page
+        if($this->is_nojs_active()){
+            // Create or update the order for the nojs template
+            $this->process_order($order_id, $crypto);
+            $this->load_blockonomics_template('nojs_checkout');
+        }else{
+            $this->load_blockonomics_template('checkout');
+        }
+    }
+
+    // Load the the crypto options template in the page
+    public function load_crypto_options_template(){
+        // Check to send the user to nojs page
+        if($this->is_nojs_active()){
+            $this->load_blockonomics_template('nojs_crypto_options');
+        }
+        else{
+            $this->load_blockonomics_template('crypto_options');
+        }
+    }
+
+    // Redirect the user to the woocommerce finish order page
+    public function redirect_finish_order($order_id){
+        $wc_order = new WC_Order($order_id);
+        wp_redirect($wc_order->get_checkout_order_received_url());
+        exit();
+    }
+
+    // Fetch all the crypto orders linked to the order id
+    public function get_all_orders_by_id($orders, $order_id){
+        if(isset($orders[$order_id])){
+            return $orders[$order_id];
+        }
+        return false;
+    }
+
+    // Check and update the crypto order or create a new order
+    public function process_order($order_id, $crypto){
+        $orders = get_option('blockonomics_orders');
+
+        $orders_by_id = $this->get_all_orders_by_id($orders, $order_id);
+        if ($orders_by_id) {
+            $this->check_for_pending_payment($orders_by_id, $order_id);
+
+            $waiting_order = $this->get_waiting_order($orders_by_id, $order_id, $crypto);
+            if ($waiting_order) {
+                $order = $waiting_order;
+            }else {
+                $order = $this->create_new_order($order_id, $crypto);
+            }
+        }else {
+            $order = $this->create_new_order($order_id, $crypto);
+        }
+        
+        $orders[$order_id][$order['address']] = $order;
+        update_option('blockonomics_orders', $orders);
+
+        $order['crypto'] = $this->getActiveCurrencies()[$crypto];
```

Comment:
This looks pretty innocuous however, it causes lots of problem with object flow and structure.  In order object crypto is a 'string' everywhere. However, suddenly you combine two objects to return consolidate data to client. This distrurbs the object structures and creates problems 
For example, see in blockonomics in some app.JS we do multiple JS calls to fetch objects. We could theoretically combine all in single call to save time that we don't do that

A simple solution is not to have this and just render *active_cryptos* div in the crypto_checkout template also. Another solution is to add another JS endpoint for active cryptos

Reply:
Fixed with 7e2bded and 7156d29